src/Microsoft.Extensions.Logging.Log4Net.AspNetCore/Entities/MessageCandidate.cs

```diff
@@ -2,29 +2,26 @@
 
 namespace Microsoft.Extensions.Logging.Log4Net.AspNetCore.Entities
 {
-    internal class MessageCandidate
+    /// <summary>
+    /// Represents a candidate for a log message that should be printed. This candidate will either be accepted or denied by the logger that is trying to print it.
+    /// </summary>
+    /// <typeparam name="TState">Type of the state that is used to format the error message.</typeparam>
+    public class MessageCandidate<TState>
     {
-        public MessageCandidate(LogLevel logLevel, object message, Exception exception)
+        public MessageCandidate(TState state, LogLevel logLevel, Exception exception, Func<TState, Exception, string> formatter)
         {
+            State = state;
             LogLevel = logLevel;
-            Message = message;
             Exception = exception;
+            Formatter = formatter;
         }
 
-        public LogLevel LogLevel { get; }
+        public TState State { get; }
```

Comment:
Doubt about this one... my fault, by ignorant.

Why to use `TState` instead object? Which is the benefit you get with this?

Reply:
1. It is more consistent with the ILogger interface from Microsoft, that uses a generic parameter for the state.
2. It's potentially faster: Objects require boxing for value types. Generics can avoid that.
3. TState assures at compile time that the `Formatter` inside the MessageCandidate is able to format the message with the provided state, because it also uses the TState parameter (Formatter and State are type compatible).